viewing_party/party.py

```diff
@@ -1,23 +1,170 @@
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if len(user_data["watched"]) == 0:
+        return 0.0
+    sum_rating = 0.0
+    for movie in user_data["watched"]:
+        sum_rating += movie["rating"]   
+    avg_rating = float(sum_rating/len(user_data["watched"]))
+    return avg_rating
+
+
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) == 0:
+        return None
+    count_genre = {}   
+    for movie in user_data["watched"]:
+        count_genre[movie["genre"]] = count_genre.get(movie["genre"],0) + 1
+    
+    max_count = 0
+    popular_genre = None
+    for gerne,count in count_genre.items():
+        if count > max_count:
+            popular_genre = gerne
+            max_count = count   
+    return popular_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_unique_movie_list = [] 
+    friends_watched_titles = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_titles.append(movie["title"])
+
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_watched_titles:
+            user_unique_movie_list.append(movie) 
+
+    return user_unique_movie_list
+
+
+def get_friends_unique_watched(user_data):
+    friends_unique_movies = []
+    user_watched_titles = []
+
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles \
+                and movie not in friends_unique_movies:
+                friends_unique_movies.append(movie)
+
+    return friends_unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies_list = []
+    user_watched_titles = get_user_watched_list(user_data)
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_watched_titles \
+                and movie["host"] in user_data["subscriptions"] \
+                and movie not in recommended_movies_list:
+                    recommended_movies_list.append(movie)
+
+    return recommended_movies_list
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    recs = []
+    if not user_data["watched"]:
+        return recs   
+    if not any(friend["watched"] for friend in user_data["friends"]):
+        return recs  # Return empty list since there are no watched movies by friends  

```

Comment:
Usually with guard clauses, we want to check to if the data we are about to work with will cause the logic following the guard clause to result in an undefined value/error. Here the loops will still work as intended and return an empty list for the result. Love to see you experimenting with `any` method!